Timer.Reset(0) should trigger the timer right away.

jameinel opened the pull request:
Clock.NewTimer(0) would trigger immediately, and a wall clock
timer.Reset() has the same behavior that Reset(<=0) causes the channel
to activate.

I came across as I'm trying to use Timer objects instead of After() calls. Because when using "WaitAdvance" each call to After() creates a new waiter, which means you need to know the exact count of passes through a loop that does:

```
for {
  select {
  case <-time.After(blah):
  case <-aDifferentCondition:
  }
}
```
In the above case, if blah is <=0, we immediately trigger the timer. But doing:
```
timer := clock.NewTimer(10*time.Second)
for {
  select {
  case <-timer.Chan():
  case <-aDifferentCondition:
    timer.Reset(0)
```

Would *not* trigger the channel to wake up after the Reset() (It does with wall clock timers.)

It is possible that we *want* this behavior, as it would let you step through it by doing:
```
 testclock.WaitAdvance(0, testing.ShortWait, 1)
```

Note that:
```
  timer := testclock.NewTimer(0)
  clock.After(0)
```
both *do* trigger immediately. Which is why I felt `Reset(0)` should as well.

jameinel committed:
Timer.Reset(0) should trigger the timer right away.

Clock.NewTimer(0) would trigger immediately, and a wall clock
timer.Reset() has the same behavior that Reset(<=0) causes the channel
to activate.

howbazaar commented on testclock/clock_test.go:
```diff
@@ -170,6 +170,94 @@ func (*clockSuite) TestNewTimerReset(c *gc.C) {
+	clock.WaitAdvance(30*time.Second, 10*time.Millisecond, 1)
+	// Now tell the goroutine to stop and start another one that *does* want to
+	// wake up
+	close(stop)
```
You need to wait for the go func to finish before moving on or there is a race with the select.

jameinel replied:
fair point. I can't get it to fail but I do see what you mean.

howbazaar commented on testclock/clock_test.go:
```diff
@@ -170,6 +170,94 @@ func (*clockSuite) TestNewTimerReset(c *gc.C) {
+		select {
+		case t := <-timer1.Chan():
+			c.Check(t0, gc.Equals, t)
+		case <-time.After(1 * time.Second):
```
one second is perhaps to short for slow architectures and noisy neighbours? Maybe?

jameinel replied:
I found that we had constants 'longWait' and 'shortWait' in the file already, so I just switched to those.
If we have to change them we can, but they should be fine. (The test runs in <1ms here, so a noisy machine would have to be 1000x slower).

howbazaar commented on testclock/clock_test.go:
```diff
@@ -170,6 +170,94 @@ func (*clockSuite) TestNewTimerReset(c *gc.C) {
+		select {
+		case <-timer2.Chan():
+			c.Errorf("timer2 should not wake up")
+		case <-time.After(50 * time.Millisecond):
```
Using 50ms wall clock time here seems flakey

jameinel replied:
I switched to 'shortWait', (which is also 50ms), because it is the normal "how long to you wait to convince yourself that something *won't* happen.

howbazaar reviewed (commented):
Seems like a simple enough change.

The tests seem a bit racy and flakey with the current times.

howbazaar commented:
testclock/clock_test.go:187:21: undefined: "github.com/juju/testing".LongWait

Perhaps you need to update the deps in this project?

jameinel committed:
Update dependencies.tsv.

Notice that we have constants in the file around longWait and shortWait,
use them. We don't need the new dependencies, but it shouldn't hurt to
use them.

jameinel reviewed (commented):
Updated dependencies and switched to the constants that are defined in the file.

jameinel commented:
$$merge$$
